cns/service/main.go

```diff
@@ -211,28 +194,59 @@ func main() {
 		}
 	}
 
-	// Set plugin options.
-	netPlugin.SetOption(acn.OptAPIServerURL, url)
+	log.Printf("stop cnm val %t", stopcnm)
+	var netPlugin network.NetPlugin
+	var ipamPlugin ipam.IpamPlugin
+
+	if !stopcnm {
+		var pluginConfig acn.PluginConfig
+		pluginConfig.Version = version
 
-	ipamPlugin.SetOption(acn.OptEnvironment, environment)
-	ipamPlugin.SetOption(acn.OptAPIServerURL, url)
-	ipamPlugin.SetOption(acn.OptIpamQueryInterval, ipamQueryInterval)
+		// Create a channel to receive unhandled errors from the plugins.
+		pluginConfig.ErrChan = make(chan error, 1)
 
-	if netPlugin != nil {
-		log.Printf("Start netplugin\n")
-		err = netPlugin.Start(&pluginConfig)
+		// Create network plugin.
+		netPlugin, err = network.NewPlugin(&pluginConfig)
 		if err != nil {
-			fmt.Printf("Failed to start network plugin, err:%v.\n", err)
+			fmt.Printf("Failed to create network plugin, err:%v.\n", err)
+			return
+		}
+
+		// Create IPAM plugin.
+		ipamPlugin, err = ipam.NewPlugin(&pluginConfig)
+		if err != nil {
+			fmt.Printf("Failed to create IPAM plugin, err:%v.\n", err)
 			return
 		}
-	}
 
-	if ipamPlugin != nil {
-		err = ipamPlugin.Start(&pluginConfig)
+		// Create the key value store.
+		pluginConfig.Store, err = store.NewJsonFileStore(platform.CNMRuntimePath + pluginName + ".json")
 		if err != nil {
-			fmt.Printf("Failed to start IPAM plugin, err:%v.\n", err)
+			fmt.Printf("Failed to create store: %v\n", err)
 			return
 		}
+
+		// Set plugin options.
+		netPlugin.SetOption(acn.OptAPIServerURL, url)
+		if netPlugin != nil {
+			log.Printf("Start netplugin\n")
+			err = netPlugin.Start(&pluginConfig)
+			if err != nil {
+				fmt.Printf("Failed to start network plugin, err:%v.\n", err)
+				return
+			}
+		}
+
+		ipamPlugin.SetOption(acn.OptEnvironment, environment)
+		ipamPlugin.SetOption(acn.OptAPIServerURL, url)
+		ipamPlugin.SetOption(acn.OptIpamQueryInterval, ipamQueryInterval)
+		if ipamPlugin != nil {
```

Comment:
Isn't ipamPlugin guaranteed to be valid at this point? We already catched err at line 217.

Reply:
not needed for netplugin either..will change both